### PR TITLE
ESD-873: [major] upgrade to sanity V3

### PR DESCRIPTION
**Pull request by joshredwards**
Jira issue: [ESD-873](https://medibank.atlassian.net/browse/ESD-873)
---

#### What does this PR do?

Rebuilds the Fieldwork plugin based off the Sanity Plugin-Kit to fall inline with V3 code standards
Converts to Typescript

[ESD-873]: https://medibank.atlassian.net/browse/ESD-873?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+[*]
+end_of_line = lf
+insert_final_newline = true
+trim_trailing_whitespace = true
+indent_style = space
+indent_size = 2
+
+[*.md]
+trim_trailing_whitespace = false
```
Comment (kprivitera) [inline .editorconfig:13]: @joshredwards is this change intentional?

File: src/components/tool/components/show-example/utils/example-to-string/index.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+type Asset = {
+  _ref: string;
+  _type: string;
+};
+
+type Block = {
+  current?: string;
+  _type: string;
+  asset?: Asset;
+  children?: any;
```
Comment (camjc) [inline src/components/tool/components/show-example/utils/example-to-string/index.ts:12]: Whats children normally typed as?

File: src/components/tool/schema-types-to-view-data/index.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  )(inputString);
+
+const schemaTypesToViewData = (types: any): KeyPair => {
+  if (!types) {
+    return [];
+  }
+
+  const countsByField = flow(
+    flatMap((type: { fields: any }) => map('name', type.fields)),
+    countBy, // For some reason the lodash/fp version of this causes everything to explode
```
Comment (camjc) [inline src/components/tool/schema-types-to-view-data/index.ts:39]: Do we need to pass it a 'by what' (identity function or something?)

File: src/components/tool/schema-types-to-view-data/index.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    reduce((currentItem, accumulator) => ((currentItem << 5) - currentItem + accumulator.charCodeAt(0)) | 0, 0),
+  )(inputString);
+
+const schemaTypesToViewData = (types: any): KeyPair => {
+  if (!types) {
+    return [];
+  }
+
+  const countsByField = flow(
+    flatMap((type: { fields: any }) => map('name', type.fields)),
```
Comment (camjc) [inline src/components/tool/schema-types-to-view-data/index.ts:38]: can it be an object?

File: src/components/tool/schema-types-to-view-data/index.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+import checkRules from './utils/check-rules';
+
+const hashCode = (inputString: string): number =>
+  flow(
+    split(''),
+    // eslint-disable-next-line no-bitwise
+    reduce((currentItem, accumulator) => ((currentItem << 5) - currentItem + accumulator.charCodeAt(0)) | 0, 0),
+  )(inputString);
+
+const schemaTypesToViewData = (types: any): KeyPair => {
```
Comment (camjc) [inline src/components/tool/schema-types-to-view-data/index.ts:32]: is it an array of the `{fields: any}` thingy below?

File: src/components/tool/index.tsx [hunk 6]
```diff
@@ -0,0 +1,123 @@
+import styled from "styled-components";
+import { useClient, useSchema } from "sanity";
+
+import schemaTypesToViewData from "./schema-types-to-view-data";
+import _ from "lodash";
```
Comment (camjc) [inline src/components/tool/index.tsx:5]: bunch of regular lodash you may not have seen, can probably run the codemod to get most of it

File: src/components/tool/components/use-times/index.tsx [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import { flatten, flow, size, values } from 'lodash/fp';
+
+import getPluralisationCharacter from './utils/get-pluralised-character';
+
+type UseTimesProps = {
+  fieldName: string;
+  groupOfItems: object;
+  items: any[];
```
Comment (camjc) [inline src/components/tool/components/use-times/index.tsx:8]: can it be an array of objects or something more defined than the any?

File: src/components/tool/components/show-example/index.tsx [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+type ShowExampleProps = {
+  fieldName: string;
+  sanityClient: any;
+  typeNames: string[];
+};
+
+const ShowExample = ({ fieldName, sanityClient, typeNames }: ShowExampleProps) => {
+  const [state, setState] = useState(null);
+  useEffect(() => {
+    const query = `*[_type in $typeNames && defined($fieldName)][0]{...}`; // TODO: ideally we get the field off the object inside the query.
```
Comment (camjc) [inline src/components/tool/components/show-example/index.tsx:15]: Should we do this then? just need a hand with the groq or something? or was this preexisting?

File: src/components/tool/components/show-example/index.tsx [hunk 9]
```diff
@@ -0,0 +1,23 @@
+import { Fragment, useEffect, useState } from 'react';
+
+import exampleToString from './utils/example-to-string';
+import { KeyPair } from '../../../../types';
+
+type ShowExampleProps = {
+  fieldName: string;
+  sanityClient: any;
```
Comment (camjc) [inline src/components/tool/components/show-example/index.tsx:8]: object perhaps?

File: package.json [hunk 10]
```diff
@@ -1,60 +1,90 @@
...
+    "typescript": "^5.3.3"
+  },
+  "peerDependencies": {
+    "react": "^18",
+    "sanity": "^3"
   },
   "engines": {
-    "node": "^16"
-  }
+    "node": ">=14"
```
Comment (camjc) [inline package.json:87]: should we consider >=16 or something?

File: package.json [hunk 11]
```diff
@@ -1,60 +1,90 @@
...
-    "prop-types": "^15.7.2",
-    "react": "^16.13.1",
-    "react-icons": "^3.10.0",
-    "styled-components": "^5.1.0"
-  },
-  "peerDependencies": {
-    "@sanity/base": "^2.0.0"
+    "react-icons": "5.0.1"
   },
   "devDependencies": {
```
Comment (camjc) [inline package.json:54]: all devdeps should be without `^`s (since they're fixed by the package lock anyway, this just makes it more obvious)

File: package.json [hunk 12]
```diff
@@ -1,60 +1,90 @@
...
+    "@sanity/incompatible-plugin": "^1.0.4",
     "lodash": "^4.17.19",
-    "prop-types": "^15.7.2",
-    "react": "^16.13.1",
-    "react-icons": "^3.10.0",
-    "styled-components": "^5.1.0"
-  },
-  "peerDependencies": {
-    "@sanity/base": "^2.0.0"
+    "react-icons": "5.0.1"
```
Comment (camjc) [inline package.json:52]: regular deps in a library/plugin should have `^`s so the installer can BYO any v5 version

File: package.json [hunk 13]
```diff
@@ -1,60 +1,90 @@
...
-  "repository": {
-    "type": "git",
-    "url": "git+https://github.com/ahmdigital/fieldwork.git"
-  },
-  "scripts": {
-    "build": "babel src -d lib --copy-files",
-    "lint": "eslint src",
-    "test": "echo \"Error: no test specified\" && exit 1"
-  },
   "keywords": [
```
Comment (camjc) [inline package.json:5]: NaBD: can report the keywords from the old one if you like

File: package.json [hunk 14]
```diff
@@ -1,60 +1,90 @@
 {
   "name": "sanity-plugin-fieldwork",
-  "version": "1.1.2",
+  "version": "1.0.0",
```
Comment (camjc) [inline package.json:3]: I'm guessing we need to keep going with the old version list? (eg maybe this is v2?)

File: package.config.ts [hunk 15]
```diff
@@ -0,0 +1,17 @@
+import { defineConfig } from '@sanity/pkg-utils'
+
+export default defineConfig({
+  legacyExports: true,
+  dist: 'dist',
+  tsconfig: 'tsconfig.dist.json',
+
+  // Remove this block to enable strict export validation
```
Comment (camjc) [inline package.config.ts:8]: Can/should we?

File: README.md [hunk 16]
```diff
@@ -1,24 +1,35 @@
-# Fieldwork
+# sanity-plugin-fieldwork
 
-Inspect your Sanity.io schema fields straight in your Studio using this tool
-
-## Some use cases
```
Comment (camjc) [inline README.md:5]: May consider bringing this section over

File: .eslintrc [hunk 17]
```diff
@@ -1,7 +1,40 @@
 {
-  "extends": "@ahmdigital/eslint-config/react-config",  
-  "parser": "@typescript-eslint/parser",
+  "root": true,
+  "env": {
+    "node": true,
+    "browser": true
+  },
+  "extends": [
+    "@ahmdigital/eslint-config/react-config",
```
Comment (camjc) [inline .eslintrc:8]: Can/should we greater leverage the eslint config repo? (notably for typescript like we do in sales repo)

**Review by camjc — DISMISSED** [review]
Looking good, left a few ideas :)

File: package.json (hunk 14 above)
Comment (camjc) [reply package.json:3]: Or actually, I guess if you merge this major this will happen automatically? :D

File: .editorconfig (hunk 1 above)
Comment (joshredwards) [reply .editorconfig:13]: It's just whats come from the `plugin-kit` setup

Comment (joshredwards) [reply .editorconfig:13]: Given V2 didn't have it, I'll remove this file.

File: src/components/tool/components/show-example/index.tsx (hunk 8 above)
Comment (joshredwards) [reply src/components/tool/components/show-example/index.tsx:15]: This was pre-existing

File: package.config.ts (hunk 15 above)
Comment (joshredwards) [reply package.config.ts:8]: No - I had to add this in order to fix the build.

File: package.json (hunk 10 above)
Comment (joshredwards) [reply package.json:87]: To be in line with the rest of our V3 changes, yes. This was generated by the plugin-kit. I've bumped to 16 now.

File: src/components/tool/components/show-example/index.tsx (hunk 9 above)
Comment (joshredwards) [reply src/components/tool/components/show-example/index.tsx]: Nope turns out Sanity had a definition!

**Review by kprivitera — DISMISSED** [review]

**Review by jheanly — APPROVED** [review]

**Review by kprivitera — APPROVED** [review]
